java-checks/src/main/java/org/sonar/java/checks/TextBlocksInComplexExpressionsCheck.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6203")
+public class TextBlocksInComplexExpressionsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MASSAGE = "Refactor this text block to a local variable or a static final field.";
+  private static final int DEFAULT_LINES_NUMBER = 5;
+  @RuleProperty(key = "MaximumNumberOfLines", 
+    description = "The maximum number of lines in a text block that can be nested into a complex expression.",
+    defaultValue = "" + DEFAULT_LINES_NUMBER)
+  private int linesNumber = DEFAULT_LINES_NUMBER;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.LAMBDA_EXPRESSION);
```

Comment:
I believe that just testing if the string is in a lambda is not enough. I consider the following example as false-positive:
```
    Supplier<String> supplier = () -> """
    <project>
      <modelVersion>4.0.0</modelVersion>
      <parent>
        <groupId>com.mycompany.app</groupId>
        <artifactId>my-app</artifactId>
        <version>1</version>
      </parent>

      <groupId>com.mycompany.app</groupId>
      <artifactId>my-module</artifactId>
      <version>1</version>
    </project>
    """;
```
Because the lambda is not used in a chain of method calls.

Reply:
We talked with @quentin-jaquier-sonarsource about such cases and decided to not report issues here, becasue it's almost the same as having text block in a variable.

Reply:
Decided to filter out all lambdas which are not parts of the argument list. Wdyt? I suppose it will cove the majority of TP and filter the majority of FP.